File: src/main/java/org/distril/beengine/world/chunk/bitarray/Pow2BitArray.java
package org.distril.beengine.world.chunk.bitarray;

import com.google.common.base.Preconditions;

public class Pow2BitArray extends BitArray {

	public Pow2BitArray(Version version, int size, int[] words) {
		super(version, size, words);
	}

	@Override
	public void set(int index, int value) {
		Preconditions.checkElementIndex(index, this.size);
		Preconditions.checkArgument(value >= 0 && value <= this.version.getMaxEntryValue(), "Max value: %s. Received value", this.version.getMaxEntryValue(), value);

		var bitIndex = index * this.version.getBits();
		var arrayIndex = bitIndex >> 5;
		var offset = bitIndex & 31;
		this.words[arrayIndex] = this.words[arrayIndex] & ~(this.version.getMaxEntryValue() << offset) | (value & this.version.getMaxEntryValue()) << offset;
	}

	@Override
	public int get(int index) {
		Preconditions.checkElementIndex(index, this.size);

		var bitIndex = index * this.version.getBits();
		var arrayIndex = bitIndex >> 5;
		var wordOffset = bitIndex & 31;
		return this.words[arrayIndex] >>> wordOffset & this.version.getMaxEntryValue();
	}
}

File: src/main/kotlin/org/distril/beengine/world/chunk/bitarray/Pow2BitArray.kt
package org.distril.beengine.world.chunk.bitarray

import org.distril.beengine.util.Utils.requireInRange

class Pow2BitArray(version: Version, size: Int, words: IntArray) : BitArray(version, size, words) {

	override fun set(index: Int, value: Int) {
		requireInRange(index, maxValue = this.size, name = "index")
		require(value >= 0 && value <= this.version.maxEntryValue) {
			"Max value: ${this.version.maxEntryValue}. Received value $value"
		}

		val bitIndex = index * this.version.bits
		val arrayIndex = bitIndex shr 5
		val offset = bitIndex and 31
		this.words[arrayIndex] = this.words[arrayIndex] and
				(this.version.maxEntryValue shl offset).inv() or
				(value and this.version.maxEntryValue shl offset)
	}

	override fun get(index: Int): Int {
		requireInRange(index, maxValue = this.size, name = "index")

		val bitIndex = index * this.version.bits
		val arrayIndex = bitIndex shr 5
		val wordOffset = bitIndex and 31
		return this.words[arrayIndex] ushr wordOffset and this.version.maxEntryValue
	}
}
